activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -36,7 +36,20 @@ def load_records_in_batch(loaders)
           end
 
           def load_records_for_keys(keys, &block)
-            scope.where(association_key_name => keys).load(&block)
+            resulting_scope = scope
+            if association_key_name.is_a?(Array)
+              array_of_values_sets = keys.dup.to_a
```

Comment:
Because of the `.pop`, I didn't want to mutate the `keys` array as other parts may expect it to be immutable. `.to_a` is to convert `Set` to an array as there is no `.pop` method on a set.

Reply:
I might be mistaken, but this looks like a case where `inject` might be helpful? 

Something like this should return a similar result without the need to `dup` and `pop`:

```ruby 
resulting_scope = keys
  .map { |values_set| scope.where(association_key_name.zip(values_set).to_h) }
  .inject(scope, &:or)
```